drivers/flash/nrf_qspi_nor.c

```diff
@@ -303,30 +310,40 @@ static void qspi_handler(nrfx_qspi_evt_t event, void *p_context)
 
 
 /* QSPI send custom command */
-static int qspi_send_cmd(const struct device *dev, const struct qspi_cmd *cmd)
+static int qspi_send_cmd(const struct device *dev, const struct qspi_cmd *cmd,
+			 bool wren)
 {
 	/* Check input parameters */
 	if (!cmd) {
 		return -EINVAL;
 	}
 
-	qspi_lock(dev);
+	size_t xfer_len = sizeof(cmd->op_code);
+
+	if (cmd->tx_buf != NULL) {
+		xfer_len += cmd->tx_buf->len;
+	}
+
+	if (cmd->rx_buf != NULL) {
+		xfer_len += cmd->rx_buf->len;
+	}
+
+	if (xfer_len > NRF_QSPI_CINSTR_LEN_9B) {
```

Comment:
Assuming bidirectional transfer which is clocked by the same clock cycles and happens in parallel: doesn't it mean that the same number of bits/bytes will be transferred both ways. OK, so if we send N bytes, including command, we expect  N bytes being clocked in.
Doesn't this mean that xfer_len should be:
`(xfer_len = sizeof(cmd->op_code) + cmd->tx_buf->len)`
and
`xfer_len <= NRF_QSPI_CINSTR_LEN_9B && cmd->rx_buf->len == cmd->tx_buf-len`
(the opcode is not stored in buffer)
?

Reply:
Yeah, that's true.  The problem is that all existing uses support either read or write, not both, and the local function isn't properly documented to indicate that you need a full buffer for both even if you only care about one-way transfers.

I had a larger rework that made the long instruction API available generally (rather than special-case RDSFDP)  and ran into that as a problem.  I'll rework this to do something more correct.

Reply:
Wow, that was really wrong.  It'd have been derefencing a null pointer on every call, since it was never true that both `tx_buf` and `rx_buf` were provided in `cmd`.

Of course, it's actually been doing that all this time.

The update should be right.